virtual-device/README.md

```diff
@@ -0,0 +1,524 @@
+# Ubuntu Core with Virtual EdgeX Device
+
+This is guide for creating an AMD64 [Ubuntu Core](https://ubuntu.com/core) OS image, preloaded with EdgeX core components and the Device Virtual service. Since the image is meant for demonstration purposes only, we will disable the security mechanisms built into EdgeX.
+
+> **Note**  
+> This guide has been tested on **Ubuntu 22.04** as the desktop OS. It may work on other Linux distributions and Ubuntu versions. 
+>
+> Some commands are executed on the desktop computer, but some others on the target Ubuntu Core system. For clarity, we use **🖥 Desktop** and **🚀 Ubuntu Core** titles for code blocks to distinguish where those commands are being executed.
+
+We will use the following tools:
+
+- [snapcraft](https://snapcraft.io/snapcraft) to manage keys in the store and build snaps
+- [YQ](https://snapcraft.io/yq) to validate YAML files and convert them to JSON
+- [ubuntu-image](https://snapcraft.io/ubuntu-image) to build the Ubuntu Core image
+- [EdgeX CLI](https://snapcraft.io/edgex-cli) to query information from EdgeX core components
+
+Install the following:
+```bash
+# 🖥 Desktop
+sudo snap install snapcraft yq
+sudo snap install ubuntu-image --classic
+```
+
+## Prepare the Gadget snap
+
+The gadget is available as a prebuilt snap in the store, however, we need to build our own to:
+- Extend the size of disk volumes to have sufficient capacity for our EdgeX snaps.
+- Add default configuration for the EdgeX snaps
+
+To build a Core20 AMD64 gadget, we use the source from [branch `20` of pc-amd64-gadget](https://github.com/snapcore/pc-amd64-gadget/tree/20).
+
+Clone the branch of the repository and enter the newly created directory:
+```bash
+# 🖥 Desktop
+git clone https://github.com/snapcore/pc-amd64-gadget.git --branch=20
+cd pc-amd64-gadget
+```
+
+Made the following modification in `gadget.yml`:
+1. Under `volumes.pc.structure`, find the item with name `ubuntu-seed` and increase its size to `1500M`. This is to make sure our snap will fit in the image.
+2. Under `volumes.pc.structure` and only if planning to use an emulator: find the item with name `ubuntu-data` and increase its size to `2G`. This is to give sufficient writable storage. When flashing on actual hardware, this volume would automatically take the whole remaining space (NEED TO VERIFY).
+3. Add the following at the top level:
+  ```yml
+  # Add default config options
+  # The child keys are unique snap IDs
+  defaults:
+    # edgexfoundry
+    AZGf0KNnh8aqdkbGATNuRuxnt1GNRKkV:
+      # Disable EdgeX security
+      # By default, the security components are enabled and external access is only
+      # permitted to authenticated users, through the API Gateway.
+      security-secret-store: "off"
```

Comment:
Should this field be `off` or `false`?
I set it to `off`, and didn't get the expected `"config": {
		"edgex-security-secret-store": false
	}` field:
```
mengyiw@ubuntu:~$ snap get edgex-device-virtual -d
{
	"app-options": true,
	"apps": {
		"device-virtual": {
			"config": {
				"service-startupmsg": "Startup message from gadget!"
			}
		}
	},
	"autostart": true
}
```

Reply:
It should be `"off"`, with double quotes. Otherwise the yaml parser converts it to false which is rejected by the config hook implementation.

Reply:
The command output is for device virtual, but the selected line related to the edgexfoundry snap. You are facing this issue because `config.edgex-security-secret-store: false` is missing from device virtual's defaults.